Implemented gfxWindowCreate

0x1F9F1 opened the pull request:
Implemented gfxWindowCreate
Added "noborder" arg - Allows fullscreen borderless window
Removed SubclassGameWindow - Its functionality is now default
Moved "nopause" byte patch - Now implemented in WndProcNew
Properly hid the regular cursor - It would randomly pop up sometimes
Enabled the "fancy" cursor whilst in windowed mode

0x1F9F1 committed:
Added disable pause patch

0x1F9F1 committed:
Added nopause arg to toggle disabling pause on focus loss

0x1F9F1 committed:
Implemented gfxWindowCreate
Added custom implentation of gfxWindowCreate - Almost identical to the original for now
Added noborder param to completely remove window border
Added operator= to MM2PtrHook

0x1F9F1 committed:
Removed the SubclassGameWindow
Implemented WndProcNew into gfxWindowCreate
Changed VTune hook types

0x1F9F1 committed:
Removed nopause byte patch - added to WndProcNew instead
Fixed window style

0x1F9F1 committed:
Enabled fancy cursor in windowed mode
Properly hid the cursor (before the loading cursor would randomly appear)

Fireboyd78 commented on src/mm2_utils.h:
```diff
@@ -171,6 +171,11 @@ class MM2PtrHook : public IMM2HookPtr
         return (*this->get())(args...);
     }
+
+    inline void operator=(TType value)
```
I actually had this implemented at one point (the '=' operator), but if you look at the `IHookPtr` implementation, it's used to set the actual pointer address, not the pointer contents. Admittedly this is due to my primitive hook system that was written for another game, and I haven't really updated it much. It wasn't broken, but it probably wasn't the best.

What do you think? I think the '=' operator should be used to reset the pointer location (e.g. 0xF00BA2 to 0xBADBA2).

Fireboyd78 commented on src/main.cpp:
```diff
@@ -119,33 +120,14 @@ MM2PtrHook<void (*)(int, LPCSTR, char *)>
-MM2PtrHook<DWORD> $__VtResumeSampling       ( 0x5C86B8, 0x5DF710, 0x5E0CC4 );
-MM2PtrHook<DWORD> $__VtPauseSampling        ( 0x5C86C8, 0x5DF724, 0x5E0CD8 );
-MM2PtrHook<BOOL> $gameClosing               ( 0x667DEC, 0x6B0150, 0x6B1708 );
+MM2PtrHook<void(__stdcall*)(void)> $__VtResumeSampling  ( 0x5C86B8, 0x5DF710, 0x5E0CC4 );
```
Technically the `__stdcall` is redundant since it has no parameters and doesn't need to clean up the stack.

Fireboyd78 commented on src/main.cpp:
```diff
@@ -433,30 +329,168 @@ BOOL __stdcall AutoDetectCallback (GUID*    lpGUID,
+            lpWindowName = lpWindowTitle;
+        }
+
+        MM2PtrHook<HWND>    hWndMain(NULL, NULL, 0x6830B8);
```
All of these PtrHook's should be moved out of the function, just to keep things neat and organized.

Fireboyd78 commented on src/main.cpp:
```diff
@@ -433,30 +329,168 @@ BOOL __stdcall AutoDetectCallback (GUID*    lpGUID,
+    }
 
-    static void setRes(int width, int height, int cdepth, int zdepth, bool detectArgs)
+    static void __cdecl gfxWindowCreate(LPCSTR lpWindowName)
```
`__cdecl` is redundant, it's the default compiler option.

Fireboyd78 commented on src/main.cpp:
```diff
@@ -433,30 +329,168 @@ BOOL __stdcall AutoDetectCallback (GUID*    lpGUID,
+    {
+        LogFile::WriteLine("Additional graphics params enabled.");
+
+        $setRes(width, height, cdepth, zdepth, datArgParser::Get("gfxArgs") ? true : detectArgs);
```
Oh hey, sorry about the "gfxArgs" thing. I was trying to debug why I had some strange flickering issues. This can be removed.

Fireboyd78 commented on src/main.cpp:
```diff
@@ -658,6 +692,96 @@ class memSafeHeapCallbackHandler
+            LogFile::Write("FAIL!\n");
+            */
+
+            if (datArgParser::Get("noautodetect"))
```
I'm thinking maybe we should change this to an int argument that works like this:

`-autodetect 1` -- Use old behavior
`-autodetect 2` -- Use new behavior (default)

```
enum AutoDetectBehavior {
    Original = 1,
    New = 2
};
```

Fireboyd78 commented:
Overall this pull request looks very nice, just need you to address those couple of things I mentioned above and we'll be good to go! 👍

0x1F9F1 committed:
Fixes for PR #4
Moves MMPtrHooks to top of file
Removed unnecessary __stcall and __cdecl specifiers
Renamed arg "noautodetect" to "oldautodetect" for better clarity
Removed operator=
